core/vm/state_manager.go

```diff
@@ -0,0 +1,131 @@
+	"incrementOvmContractNonce(address)":            incrementOvmContractNonce,
+	"getCodeContractBytecode(address)":              getCodeContractBytecode,
+	"getCodeContractHash(address)":                  getCodeContractHash,
+	"getCodeContractAddressFromOvmAddress(address)": getCodeContractAddress,
+	"associateCodeContract(address,address)":        associateCodeContract,
+	"registerCreatedContract(address)":              registerCreatedContract,
+}
+var methodIds map[[4]byte]stateManagerFunction
+
+func init() {
+	methodIds = make(map[[4]byte]stateManagerFunction, len(funcs))
+	for methodSignature, f := range funcs {
+		methodIds[methodSignatureToMethodID(methodSignature)] = f
+	}
+}
+
+func methodSignatureToMethodID(methodSignature string) [4]byte {
+	var methodID [4]byte
+	copy(methodID[:], crypto.Keccak256([]byte(methodSignature)))
+	return methodID
+}
+
+func callStateManager(input []byte, evm *EVM, contract *Contract) (ret []byte, err error) {
+	var methodID [4]byte
+	if len(input) == 0 {
+		return nil, nil
+	}
+	copy(methodID[:], input[:4])
+	ret, err = methodIds[methodID](evm, contract, input)
+	return ret, err
+}
+
+func setStorage(evm *EVM, contract *Contract, input []byte) (ret []byte, err error) {
+	address := common.BytesToAddress(input[4:36])
+	key := common.BytesToHash(input[36:68])
+	val := common.BytesToHash(input[68:100])
+	log.Debug("[State Mgr] Setting storage.", "Contract address:", hex.EncodeToString(address.Bytes()), "key:", hex.EncodeToString(key.Bytes()), "val:", hex.EncodeToString(val.Bytes()))
+	evm.StateDB.SetState(address, key, val)
+	return nil, nil
+}
+
+func getStorage(evm *EVM, contract *Contract, input []byte) (ret []byte, err error) {
+	address := common.BytesToAddress(input[4:36])
+	key := common.BytesToHash(input[36:68])
+	val := evm.StateDB.GetState(address, key)
+	log.Debug("[State Mgr] Getting storage.", "Contract address:", hex.EncodeToString(address.Bytes()), "key:", hex.EncodeToString(key.Bytes()), "val:", hex.EncodeToString(val.Bytes()))
+	return val.Bytes(), nil
+}
+
+func getCodeContractBytecode(evm *EVM, contract *Contract, input []byte) (ret []byte, err error) {
+	address := common.BytesToAddress(input[4:36])
+	code := evm.StateDB.GetCode(address)
+	log.Debug("[State Mgr] Getting Bytecode.", " Contract address:", hex.EncodeToString(address.Bytes()), "Code:", hex.EncodeToString(code))
+	return simpleAbiEncode(code), nil
+}
+
+func getCodeContractHash(evm *EVM, contract *Contract, input []byte) (ret []byte, err error) {
+	address := common.BytesToAddress(input[4:36])
+	codeHash := evm.StateDB.GetCodeHash(address)
+	log.Debug("[State Mgr] Getting Code Hash.", " Contract address:", hex.EncodeToString(address.Bytes()), "Code hash:", hex.EncodeToString(codeHash.Bytes()))
+	return codeHash.Bytes(), nil
+}
+
+func associateCodeContract(evm *EVM, contract *Contract, input []byte) (ret []byte, err error) {
+	log.Debug("[State Mgr] Associating code contract")
+	return []byte{}, nil
+}
```

Comment:
Ah this is used on L1 just not in L2. We still need to gas meter it / handle requests for it so that we run identical L1 & L2 code.

also re. your question

> In what line do we enforce that the stateManager can only be called by the EM?

In the safety checker!